packages/postgres-database/tests/projects/test_projects_version_control.py

```diff
@@ -0,0 +1,285 @@
+        assert branch
+        assert branch.name == "main"
+        assert branch.head_commit_id is None, "First commit"
+
+        # create commit
+        commits_orm = CommitsOrm(conn)
+        commit_id = await commits_orm.insert(
+            repo_id=repo.id,
+            parent_commit_id=branch.head_commit_id,
+            snapshot_checksum=snapshot_checksum,
+            message="first commit",
+        )
+        assert commit_id
+        assert isinstance(commit_id, int)
+
+        # update branch head
+        await branches_orm.update(head_commit_id=commit_id)
+
+        # update checksum cache
+        await repo_orm.update(project_checksum=snapshot_checksum)
+
+    # log history
+    commits = await commits_orm.fetchall()
+    assert len(commits) == 1
+    assert commits[0].id == commit_id
+
+    # tag
+    tag_orm = TagsOrm(conn)
+    tag_id: Optional[int] = await tag_orm.insert(
+        repo_id=repo.id,
+        commit_id=commit_id,
+        name="v1",
+    )
+    assert tag_id is not None
+
+    tag = await tag_orm.fetch(rowid=tag_id)
+    assert tag
+    assert tag.name == "v1"
+
+    ############# NEW COMMIT #####################
+
+    # user add some changes
+    repo = await repo_orm.fetch()
+    assert repo
+
+    project_orm.pin(uuid=repo.project_uuid)
+    assert project_orm.is_pinned()
+
+    await project_orm.update(
+        workbench={
+            "node": {
+                "input": 3,
+            }
+        }
+    )
+
+    project_wc = await project_orm.fetch("workbench ui")
+    assert project_wc
+    assert project.workbench != project_wc.workbench
+
+    # get HEAD = repo.branch_id -> .head_commit_id
+    head = await heads_orm.fetch()
+    assert head
+    branch = await branches_orm.fetch("head_commit_id", rowid=head.head_branch_id)
+    assert branch
+    # TODO: get subquery ... and compose
+    head_commit = await commits_orm.fetch(rowid=branch.head_commit_id)
+    assert head_commit
+
+    # compare checksums between wc and HEAD
+    checksum = eval_checksum(project_wc.workbench)
+    assert head_commit.snapshot_checksum != checksum
+
+    # updates wc checksum cache
+    await repo_orm.update(project_checksum=checksum)
+
+    # take snapshot = add & commit
+    async with conn.begin():
+        snapshot_uuid: str = await add_snapshot(project_wc, checksum, repo, conn)
+
+        commit_id = await commits_orm.insert(
+            repo_id=head_commit.repo_id,
+            parent_commit_id=head_commit.id,
+            snapshot_checksum=checksum,
+            message="second commit",
+        )
+        assert commit_id
+        assert isinstance(commit_id, int)
+
+        # update branch head
+        await branches_orm.update(head_commit_id=commit_id)
+
+    # log history
+    commits = await commits_orm.fetchall()
+    assert len(commits) == 2
+    assert commits[1].id == commit_id
+
+    ############# CHECKOUT TO TAG #####################
+
+
+@pytest.mark.skip(reason="DEV")
+def test_concurrency():
```

Comment:
hmm, this one will be funny I gather.

Reply:
yes that is a hard one ...therefore i do not want to forget